// sails-js/lib/cjs/transaction-builder.cjs
var api = require('@gear-js/api');
var payloadMethod = require('./util/lib/payload-method.cjs');
var consts = require('./consts.cjs');
var utils = require('./utils.cjs');

class TransactionBuilder {
    _api;
    _registry;
    _responseType;
    _onProgramCreated;
    _account;
    _signerOptions;
    _tx;
    _voucher;
    _gasInfo;
    programId;
    constructor(_api, _registry, extrinsic, payload, payloadType, _responseType, _programIdOrCodeOrCodeId, _onProgramCreated) {
        this._api = _api;
        this._registry = _registry;
        this._responseType = _responseType;
        this._onProgramCreated = _onProgramCreated;
        const _payload = this._registry.createType(payloadType, payload);
        switch (extrinsic) {
            case 'send_message': {
                this.programId = _programIdOrCodeOrCodeId;
                this._tx = this._api.message.send({
                    destination: this.programId,
                    gasLimit: 0,
                    payload: _payload.toU8a(),
                    value: 0,
                });
                break;
            }
            case 'upload_program': {
                const { programId, extrinsic } = this._api.program.upload({
                    code: _programIdOrCodeOrCodeId,
                    gasLimit: 0,
                    initPayload: _payload.toU8a(),
                });
                this.programId = programId;
                this._tx = extrinsic;
                break;
            }
            case 'create_program': {
                const { programId, extrinsic } = this._api.program.create({
                    codeId: _programIdOrCodeOrCodeId,
                    gasLimit: 0,
                    initPayload: _payload.toU8a(),
                });
                this.programId = programId;
                this._tx = extrinsic;
                break;
            }
        }
    }
    _getGas(gas, increaseGas) {
        if (increaseGas === 0)
            return gas;
        if (increaseGas < 0 || increaseGas > 100)
            throw new Error('Invalid increaseGas value (0-100)');
        return this._registry.createType('u64', gas.add(gas.muln(increaseGas / 100)));
    }
    _getValue(value) {
        return this._registry.createType('u128', value);
    }
    _setTxArg(index, value) {
        const args = this._tx.args.map((arg, i) => (i === index ? value : arg));
        switch (this._tx.method.method) {
            case 'uploadProgram': {
                this._tx = this._api.tx.gear.uploadProgram(...args);
                break;
            }
            case 'createProgram': {
                this._tx = this._api.tx.gear.createProgram(...args);
                break;
            }
            case 'sendMessage': {
                this._tx = this._api.tx.gear.sendMessage(...args);
                break;
            }
        }
    }
    /** ## Get submittable extrinsic */
    get extrinsic() {
        return this._tx;
    }
    /** ## Get payload of the transaction */
    get payload() {
        return this._tx.args[0].toHex();
    }
    /**
     * ## Calculate gas for transaction
     * @param allowOtherPanics Allow panics in other contracts to be triggered (default: false)
     * @param increaseGas Increase the gas limit by a percentage from 0 to 100 (default: 0)
     * @returns
     */
    async calculateGas(allowOtherPanics = false, increaseGas = 0) {
        const source = this._account
            ? api.decodeAddress(typeof this._account === 'string' ? this._account : this._account.address)
            : consts.ZERO_ADDRESS;
        let gas;
        switch (this._tx.method.method) {
            case 'uploadProgram': {
                gas = await this._api.program.calculateGas.initUpload(source, this._tx.args[0].toHex(), this._tx.args[2].toHex(), this._tx.args[4], allowOtherPanics);
                this._setTxArg(3, this._getGas(gas.min_limit, increaseGas));
                break;
            }
            case 'createProgram': {
                gas = await this._api.program.calculateGas.initCreate(source, this._tx.args[0].toHex(), this._tx.args[2].toHex(), this._tx.args[4], allowOtherPanics);
                this._setTxArg(3, this._getGas(gas.min_limit, increaseGas));
                break;
            }
            case 'sendMessage': {
                gas = await this._api.program.calculateGas.handle(source, this._tx.args[0].toHex(), this._tx.args[1].toHex(), this._tx.args[3], allowOtherPanics);
                this._setTxArg(2, this._getGas(gas.min_limit, increaseGas));
                break;
            }
            default: {
                throw new Error('Unknown extrinsic');
            }
        }
        this._gasInfo = gas;
        return this;
    }
    /**
     * ## Set account for transaction
     * @param account
     * @param signerOptions
     */
    withAccount(account, signerOptions) {
        this._account = account;
        if (signerOptions) {
            this._signerOptions = signerOptions;
        }
        return this;
    }
    /**
     * ## Set value for transaction
     * @param value
     */
    withValue(value) {
        switch (this._tx.method.method) {
            case 'uploadProgram':
            case 'createProgram': {
                this._setTxArg(4, this._getValue(value));
                break;
            }
            case 'sendMessage': {
                this._setTxArg(3, this._getValue(value));
                break;
            }
            default: {
                throw new Error('Unknown extrinsic');
            }
        }
        return this;
    }
    /**
     * ## Set gas for transaction
     * @param gas
     */
    withGas(gas) {
        switch (this._tx.method.method) {
            case 'uploadProgram':
            case 'createProgram': {
                this._setTxArg(3, this._registry.createType('u64', gas));
                break;
            }
            case 'sendMessage': {
                this._setTxArg(2, this._registry.createType('u64', gas));
                break;
            }
            default: {
                throw new Error('Unknown extrinsic');
            }
        }
        return this;
    }
    /**
     * ## Use voucher for transaction
     * @param id Voucher id
     */
    withVoucher(id) {
        if (this._tx.method.method !== 'sendMessage') {
            throw new Error('Voucher can be used only with sendMessage extrinsics');
        }
        this._voucher = id;
        return this;
    }
    /**
     * ## Get transaction fee
     */
    async transactionFee() {
        if (!this._account) {
            throw new Error('Account is required. Use withAccount() method to set account.');
        }
        const info = await this._tx.paymentInfo(this._account, this._signerOptions);
        return info.partialFee.toBigInt();
    }
    /**
     * ## Sign and send transaction
     */
    async signAndSend() {
        if (!this._account) {
            throw new Error('Account is required. Use withAccount() method to set account.');
        }
        if (this._voucher) {
            const callParams = { SendMessage: this._tx };
            this._tx = this._api.voucher.call(this._voucher, callParams);
        }
        let resolveFinalized;
        const isFinalized = new Promise((resolve) => {
            resolveFinalized = resolve;
        });
        const { msgId, blockHash, programId } = await new Promise((resolve, reject) => this._tx
            .signAndSend(this._account, this._signerOptions, ({ events, status }) => {
            if (status.isInBlock) {
                let msgId;
                let programId;
                for (const { event } of events) {
                    const { method, section, data } = event;
                    if (section == 'gear') {
                        if (method === 'MessageQueued') {
                            msgId = data.id.toHex();
                        }
                        else if (method == 'ProgramChanged') {
                            programId = data.id.toHex();
                        }
                    }
                    else if (method === 'ExtrinsicSuccess') {
                        resolve({ msgId, blockHash: status.asInBlock.toHex(), programId });
                    }
                    else if (method === 'ExtrinsicFailed') {
                        reject(this._api.getExtrinsicFailedError(event));
                    }
                }
            }
            else if (status.isFinalized) {
                resolveFinalized(true);
            }
        })
            .catch((error) => {
            reject(error.message);
        }));
        if (this._onProgramCreated && programId) {
            await this._onProgramCreated(programId);
        }
        return {
            msgId,
            blockHash,
            txHash: this._tx.hash.toHex(),
            isFinalized,
            response: async (rawResult = false) => {
                const { data: { message: { payload, details }, }, } = await this._api.message.getReplyEvent(this.programId, msgId, blockHash);
                utils.throwOnErrorReply(details.unwrap().code, payload, this._api.specVersion, this._registry);
                if (rawResult) {
                    return payload.toHex();
                }
                // prettier-ignore
                return this._registry.createType(`(String, String, ${this._responseType})`, payload)[2][payloadMethod.getPayloadMethod(this._responseType)]();
            },
        };
    }
    get gasInfo() {
        return this._gasInfo;
    }
}

exports.TransactionBuilder = TransactionBuilder;
